packages/grafana-ui/src/components/uPlot/context.ts

```diff
@@ -0,0 +1,142 @@
+import React, { useCallback, useContext } from 'react';
+import uPlot from 'uplot';
+import { PlotPlugin } from './types';
+import { DataFrame, Field, FieldConfig } from '@grafana/data';
+
+interface PlotCanvasContextType {
+  // canvas size css pxs
+  width: number;
+  height: number;
+  // plotting area bbox, css pxs
+  plot: {
+    width: number;
+    height: number;
+    top: number;
+    left: number;
+  };
+}
+
+interface PlotContextType {
+  u: uPlot;
+  series: uPlot.Series;
+  canvas: PlotCanvasContextType;
+}
+
+interface PlotDataContextType {
+  data: DataFrame;
+}
+
+type PlotPluginsContextType = {
+  registerPlugin: (plugin: PlotPlugin) => () => void;
+};
+
+export const PlotContext = React.createContext<PlotContextType | null>(null);
+export const PlotPluginsContext = React.createContext<PlotPluginsContextType>(null);
+export const PlotDataContext = React.createContext<PlotDataContextType | null>(null);
+
+// Exposes uPlot instance and bounding box of the entire canvas and plot area
+export const usePlotContext = (): PlotContextType | null => {
+  return useContext<PlotContextType>(PlotContext);
+};
+
+// Exposes API for registering uPlot plugins
+export const usePlotPluginContext = (): PlotPluginsContextType => {
+  return useContext(PlotPluginsContext);
+};
+
+interface PlotDataAPI {
+  /** Data frame passed to graph, x-axis aligned */
+  data: DataFrame;
+  /** Returns field by index */
+  getField: (idx: number) => Field;
+  /** Returns x-axis fields */
+  getXAxisFields: () => Field[];
+  /** Returns x-axis fields */
+  getYAxisFields: () => Field[];
+  /** Returns field value by field and value index */
+  getFieldValue: (fieldIdx: number, rowIdx: number) => any;
+  /** Returns field config by field index */
+  getFieldConfig: (fieldIdx: number) => FieldConfig;
+}
```

Comment:
Why is this all necessary?  If you are passing in DataFrame, i would expect:

```
interface PlotDataAPI {
  /** Data frame passed to graph, x-axis aligned */
  data: DataFrame;
  getXAxisFields: () => number[];
  getYAxisFields: () => number[];
```

but even that makes me nervous... very soon you are going to want other types of dimenstions as well!
```
{
 x:[0],
 y:[2,3],
 size:[4],
}
```
Actually thinking about it... this is where the single frame will fall over :cry:

Reply:
@ryantxu this API is to expose original data frame rather than particular values. It's not used for the uPlot to build the model, but for plugins to get access to the fields within input data frame. See [TooltipPlugin](https://github.com/grafana/grafana/pull/27396/files#diff-5d849ca51f1c1f3d960849f4d14680ff) for an example. I think exposing the entire fields makes it more flexible to use the data in the plugins.

I'm OK with removing the X/Y specific functions for now and rely more on the `data` directly in the plugins.